File: 2015/python/09.py
# Advent of Code 2015, day 9
# (c) aichingert

import itertools

dist = {}
cities = set()

def solve(dist,cities,part):
    if part: m = 99999999
    else: m = 0

    for p in itertools.permutations(cities):
        cur = 0
        for i in range(1, len(p)):
            cur += dist[p[i-1],p[i]]
        if part:
            m = min(m, cur)
        else:
            m = max(m, cur)
    return m

for l in open("../input/09").read().split('\n'):
    vls = l.split(' ')
    dist[(vls[0],vls[2])] = int(vls[4])
    dist[(vls[2],vls[0])] = int(vls[4])
    cities.add(vls[0])
    cities.add(vls[2])

print("Part 1:",solve(dist,cities,True))
print("Part 2:",solve(dist,cities,False))

File: 2015/julia/09.jl
# Advent of Code 2015, day 9
# (c) aichingert

module Day9

using Combinatorics

function solve(dist, cities) 
	partone = typemax(Int32)
	parttwo = 0

	for perm in collect(permutations(cities))
		cur = 0
		for i in 1:length(perm)-1 cur += dist[(perm[i],perm[i+1])] end
		partone = min(partone, cur) 
		parttwo = max(parttwo, cur)
	end

	(partone,parttwo)
end

dist = Dict()
cities = []

for l in readlines("../input/09")
	locs,distance = split(l, " = ")
	from,to = split(locs, " to ")
	push!(dist, (from,to) => parse(Int32, distance))
	push!(dist, (to,from) => parse(Int32, distance))
	if !(from in cities) push!(cities, from) end
	if !(to in cities) push!(cities, to) end
end

partone,parttwo = solve(dist, cities)

println("Part 1: $(partone)")
println("Part 2: $(parttwo)")

end
